### DISCUSSION TITLE
External datasource or Replicate Database

### DISCUSSION THREAD

**Discussion by askrabz** [supabase/supabase#18840] [Questions]
Hi,
We really enjoy using Supabase but we seem to have hit some form of roadblock with fetching the data outside Supabase. Therefore, I was wondering if any of the below is possible: 

1. Set up a DB in superbase that is a part of replication from an External Postgres database (primary database) into a Supabase database (secondary database) ?  OR
2. Establish a Direct Connection from Supabase with an external Postgres DB for the purpose of fetching or reading data.

The reason for my question is that I'm building a frontend application using Supabase as BaaS and I have a blend of API and Postgres database from our back-end core system. Some of the API does not have the full capability of fetching all the necessary data, therefore, we are required to connect directly with the database (outside Supabase) and fetch the data. It would be ideal if we could have a replica of the external database inside Supebase that would solve our issue or else anyway to direct connect from Supabase. 

Any suggestion would be appreciated in resolving this issue.

Thanks

**Comment by khalidibrahimio** [comment]
Hi, we have this same exact query: we would really appreciate any guidance/input on how to use our external postgres database(s) in Supabase. The most obvious thing we thought of as well was replication but logical replication doesn't seem to be supported since the `CREATE SUBSCRIPTION` command is not supported even for the postgres superuser according to [this](https://supabase.com/docs/guides/database/postgres/roles-superuser)
Tagging @GaryAustin1 if you can help with this or point us to anyone who could 🙏🏼 🙏🏼

**Comment by Rubenburdin** [comment]
You and @khalidibrahimio are hitting the same wall others have run into. Supabase can publish changes outward via logical replication, but subscribing to an external Postgres as a replica target isn't supported. The CREATE SUBSCRIPTION command needs superuser level privileges that Supabase doesn't grant, even to the postgres role, because of how they manage the underlying infrastructure.
For option 2 (direct connection to external Postgres), you can actually do this with the postgres_fdw extension. It lets you create foreign tables in Supabase that point to your external database, and then query them like local tables. Your frontend would still hit Supabase, but under the hood Supabase fetches from your backend system. The catch is latency, every query has to reach out to your external db in real time, and you lose the ability to use RLS on those foreign tables in a meaningful way. Docs here (https://supabase.com/docs/guides/database/extensions/wrappers/overview).

If your data doesn't change constantly and you can tolerate some lag, you could combine postgres_fdw with pg_cron to periodically copy data into local Supabase tables. That way your frontend queries hit local data with full RLS support, but you're doing scheduled syncs rather than true replication.

For something closer to real time without building the plumbing yourself, external sync tools exist for this. Full disclosure, I'm the CEO of [Stacksync ](https://www.stacksync.com/)and we do Postgres to Supabase sync with change data capture, so updates flow in as they happen rather than on a schedule. Could be worth exploring if the FDW approach feels too manual for your use case.
